### ISSUE TITLE
NeXTSTEP installation failure using SCSI

### ISSUE THREAD

**Issue by win2kgamer** [86Box/86Box#356]
When installing any version of NeXTSTEP using a SCSI hard disk (using either an Adaptec or BusLogic SCSI adapter), the installation fails with "DISK INITIALIZATION FAILED." NeXTSTEP 3.1 is a little more verbose and displays "Writing /usr/standalone/i386/boot1" before the initialization failure message. Using an IDE hard disk with a SCSI CD-ROM does work.

86Box configuration:
Machine: AMI ALi 1429
CPU: Intel 486DX2/66
Memory: 16MB
Video: [VLB] Phoenix S3c805 2MB
Mouse: Microsoft Serial Mouse
SCSI Controller: Adaptec AHA-1542C (I/O 0x330, IRQ 11, DMA 5, BIOS 0xC800)
No sound card or network controller
I've tried several hard disk sizes making sure it wasn't an OS limitation
Hard disk on SCSI ID 0
SCSI CD-ROM on SCSI ID 6

**Comment by TC1995** [comment]
in 3.2 and 3.3, does it also happens with the 53c8xx?

**Comment by ArtiomWin** [comment]
I confirm that the bug affects not only NEXTstep - other systems also don't see HDDs connected to SCSI.

**Comment by OBattler** [comment]
Which other systems specifically?

Also, I have this fixed on my end (it was a problem with the MODE SENSE pages that report the drive's geometry).

**Comment by ArtiomWin** [comment]
SCO Open Desktop 2.0 - when both HDD and CD-ROM are connected to SCSI, Installation Query Manager fails to start, while when HDD is connected to IDE, installation goes on... until the moment of looking up for CD-ROM (which was connected to SCSI, of course, since that version doesn't supported IDE CD-ROMs).
![20190607_212013](https://user-images.githubusercontent.com/31134876/59126377-9bcf3780-896d-11e9-813d-5f10808c3ff0.png)
![20190607_214249](https://user-images.githubusercontent.com/31134876/59126378-9bcf3780-896d-11e9-8295-8d534ac6b596.png)

**Comment by ArtiomWin** [comment]
Here's config and nvr folder.
[cfg+nvr.zip](https://github.com/86Box/86Box/files/3267245/cfg%2Bnvr.zip)

**Comment by edanidzerda** [comment]
Hello!  I am new to 86box and OpenStep, but it has been very fun to experiment with.  I love the PC Beep on startup!! 💯  

I've been testing with version v6.0 [build 9001] [4fef696a4] on Windows and Mac OS.  I've been able to reproduce the original posters problem on both platforms.

However, I built from the latest master branch to try to debug, and it worked!!!  I need to do a little more testing, but I was able to initialize a NextDisk with SCSI. I wanted to see if I could install and boot OpenStep from CD without any additional drivers, IE, with a SCSI CDROM and SCSI disk.

So I'm really just commenting in case someone else is trying to do the same thing.  It was still broken in 6.0, but it looks like it is fixed in 7.0.

My testing with OpenStep 4.2.

Machine:

- [ALi ALADDiN V] ASUS P5A
- Intel Pentium MMX / 233 MHz
- Memory: 64MB
- Video: [PCI] Cirrus Logic GD5446
- Storage:  
  - IDE Disk 1: 2.0GB
  - SCSI Disk 1: 2.0GB
  - SCSI Controller: BusLogic BT-958D
  - CDROM: 8x 86Box 86B_CD 3.50 (ATAPI)
  - Floppy:  (internal) 3.5" 1.44M
- Sound: [ISA16] Sound Blaster 16 PNP
- Network: [PCI] AMD PCnet-PCI II (SLiRP)

This is the error when initializing a SCSI disk in 6.0 release:

```
disk> init
DESTROYS ALL EXISTING DISK DATA -- really initialize? y
enter host name: myhost
enter disk label: scsi1
writing disk label
Boot block is "/usr/standalone/i386/boot", ok? y
Writing /usr/standalone/i386/boot
Writing /usr/standalone/i386/boot1
creating new filesystem on /dev/rsd1a
/usr/etc/newfs -n -v /dev/rsd1a
/etc/mkfs /dev/rsd1a 2046584 32 16 8192 1024 16 10 60 4096 t
Warning: 392 sector(s) in last cylinder unallocated
/dev/rsd1a:     2046584 sectors in 3998 cylinders of 16 tracks, 32 sectors
        2095.7Mb in 250 cyl groups (16 c/g, 8.39Mb/g, 1984 i/g)
super-block backups (for fsck -b#) at:write error: 32
wtfs: I/O error
/usr/etc/newfs /dev/rsd1a failed (status 1)
disk> ␦
Stopped
myhost:5# tail /usr/adm/mess
/usr/adm/mess: No such file or directory
myhost:6# tail /usr/adm/messages
Apr 18 04:07:26 myhost mach: sd1: TRANSFER COUNT ERROR.  Expected = 253952 Received 0; Retrying.
Apr 18 04:07:26 myhost mach:    target:0 lun:0 op:Write block:384 blockCount:496
Apr 18 04:07:26 myhost mach: sd1: TRANSFER COUNT ERROR.  Expected = 253952 Received 0; Retrying.
Apr 18 04:07:26 myhost mach:    target:0 lun:0 op:Write block:384 blockCount:496
Apr 18 04:07:26 myhost mach: sd1: TRANSFER COUNT ERROR.  Expected = 253952 Received 0; Retrying.
```

Let me know if there's any other testing I could help with on this, been really enjoying 86box!!

**Comment by OBattler** [comment]
It was most likely this commit: https://github.com/86Box/86Box/commit/74818ce82b6bc1646b6e929982f76a9d9518f4bd#diff-012d063417f9381bbb513f64c79c66fcf7cc1ba07f7c6c298887dcb13782ab51 . This was causing erroneously 0-byte transfers with some commands.

**Comment by OBattler** [comment]
That would correspond to build 9470: https://ci.86box.net/job/86Box/9470/ .

**Comment by edanidzerda** [comment]
Thanks a lot!  I thought it would be fixed in 6.0 since it the issue was closed so long ago.

Happy to report that on my Macbook Pro with latest source (7ff8eb7) I was able to install OpenStep 4.2 using the SCSI controller `[PCI] AMD 53c974`. It worked from the floppy installer through to a full boot up.  I also tested the `BusLogic 958T` for copying between disks, and at least 8K block sizes worked in OpenStep using dd.  

Thanks again!
